src/DiscreteSystem.ts

```diff
@@ -154,6 +162,49 @@ export class DiscreteSystem<TShared, TInternal> implements System {
         this._time = time;
     }
 
+    /**
+     * Runs the system from its current time to a series of times given by the parameter times
+     * and returns state values for all particles at each of these times.
+     * @param times The times to run to and return state for. Must be in increasing order, with no value less than the
+     * current time.
+     * @param stateElementIndices Indices of the state elements to return in the result. If an empty array is provided,
+     * all values are returned.
+     */
+    public simulate(times: number[], stateElementIndices: number[] = []): SystemSimulateResult {
+        checkTimes(times, this._time);
+        if (stateElementIndices.length) {
+            checkIndicesForMax("State Element", stateElementIndices, this._state.nStateElements - 1);
+        }
+
+        // Allow case where first time is current time - no progression needed for first result set
+        let preRunSaveValues = times[0] === this._time;
+
+        const stateIndicesToReturn = stateElementIndices.length
+            ? stateElementIndices
+            : [...Array(this._state.nStateElements).keys()];
+
+        const result = new SystemSimulateResult(
+            this._nGroups,
+            this._nParticles,
+            stateIndicesToReturn.length,
+            times.length
+        );
+        times.forEach((t, iTime) => {
+            if (preRunSaveValues) {
+                preRunSaveValues = false;
+            } else {
+                this.runToTime(t);
+            }
```

Comment:
This would be tidier done in `runToTime` so that it's a noop if `t === current time`